### ISSUE TITLE
Debug TSP example on Julia 1.7

### ISSUE THREAD

**Issue by odow** [jump-dev/JuMP.jl#2862]
Per @jd-foster on Gitter:

<img width="913" alt="image" src="https://user-images.githubusercontent.com/8177701/152661740-e6e49098-d187-4384-8988-34086ebc2bc6.png">

**Comment by danielschermer** [comment]
I believe that this might be related to the underlying Optimizer and how it handles lazy constraints (or the interface in between). Using Gurobi as an alternative Optimizer, no duplicate cycles are detected (and the lazy method scales much better). I checked which cycles are detected through:

`println("Node $(GLPK.glp_ios_curr_node(cb_data.tree)): Found cycle of length $(length(cycle)) $cycle")`

Using GLPK, for the value of `n=22` (for smaller values of `n` it seems to work fine, i.e., no duplicate cycles) the iterated model finds the following cycles:

```
Found cycle of length 3: [16, 10, 8]
Found cycle of length 4: [1, 14, 11, 2]
Found cycle of length 4: [7, 21, 13, 4]
```

The lazy constraint methods finds the following cycles. Note that identical cycles are detected at different nodes in the tree.

```
Node 5: Found cycle of length 3 [16, 10, 8] <- as in iterated
Node 18: Found cycle of length 3 [12, 22, 3]
Node 22: Found cycle of length 3 [1, 14, 2]
Node 22: Found cycle of length 4 [7, 21, 13, 4] <- as in iterated
Node 22: Found cycle of length 3 [17, 18, 9]
Node 22: Found cycle of length 4 [5, 20, 6, 15]
Node 22: Found cycle of length 5 [1, 14, 11, 19, 2]
Node 30: Found cycle of length 3 [1, 11, 2]
Node 30: Found cycle of length 3 [16, 10, 8] <- should not appear again
Node 30: Found cycle of length 4 [7, 21, 13, 4] <- should not appear again
Node 31: Found cycle of length 3 [16, 10, 8] <- should not appear again
Node 43: Found cycle of length 3 [12, 22, 3] <- should not appear again
Node 48: Found cycle of length 4 [7, 21, 13, 4] <- should not appear again
Node 27: Found cycle of length 3 [16, 10, 8] <- should not appear again
Node 48: Found cycle of length 4 [7, 21, 13, 4] <- should not appear again
Node 29: Found cycle of length 3 [16, 10, 8] <- should not appear again
Node 25: Found cycle of length 4 [1, 14, 11, 2] <- as in iterated
Node 25: Found cycle of length 4 [7, 21, 13, 4] <- should not appear again
```

Unfortunately, I am not very familiar with GLPK, but it seems to me that this might just be how it is designed and intended to function. If we refer to the [pseudocode on page 120 and 121 of the reference manual](https://cse.iitkgp.ac.in/~abhij/course/lab/CompLab-I/Autumn19/glpk.pdf) and look at step 4 ("Adding lazy constraints"), it appears to me that lazy constraints are only ever locally to currently investigated nodes in the tree. Not globally, to all further nodes which might be inspected. This might be, why the lazy constraint method becomes much slower than the iterated method, as we increase `n`.

**Comment by danielschermer** [comment]
I did some further investigating. The branch-and-cut tree looks as follows (for `n=22` and the fixed seed). Be careful: As indicated by the prime (') some reference numbers are used repeatedly by GLPK; most likely if a node has been previously fathomed.

What we can see is that cycles (e.g., `[16,10,8]`) are only detected if they have not been detected before by a parent (or any other  ancestor), which might explain the weak performance: Even though the cycle `[16,10,8]` is found (quite early) at Node 5, much time is spent at descendants of Node 3, finding the very same cycle.

Of course, it would be interesting to see which branching action is taken at which node but I have not managed to do this yet...

```
Node 1 (Root)
    Node 2
        Node 4
        Node 5 [16,10,8]
            Node 6
                Node 8
                    Node 10
                        Node 12
                            Node 14
                                Node 16
                                    Node 18 [12,22,3]
                                        Node 20
                                            Node 22 [1, 14, 2], [7, 21, 13, 4], [17, 18, 9], [5, 20, 6, 15], [1, 14, 11, 19, 2]
                                                Node 26
                                                    Node 27
                                                    Node 28
                                                        Node 29
                                                Node 25 [1,14,11,2]                
    Node 3
        Node 22'
            Node 26'
                Node 27' [16, 10, 8]
                    Node 48' [7, 21, 13, 4]
                Node 28' 
                    Node 29 [16, 10, 8]
                    Node 30 [1, 11, 2]
            Node 25 [1, 14, 11, 2], [7, 21, 13, 4]
        Node 24
            Node 30' [16, 10, 8], [7, 21, 13, 4]
            Node 31 [16, 10, 8]
                Node 30''
                    Node 33
                        Node 35
                            Node 37
                                Node 39
                                    Node 41
                                        Node 43 [12, 22, 3]
                                            Node 45
                                                Node 48 [7, 21, 13, 4]
```

Maybe the best solution is to reduce `n` in the Documentation (e.g., `n=20`) and add a warning?

**Comment by mlubin** [comment]
Nice investigative work. More evidence in favor of your conclusion that GLPK can't apply lazy constraints globally is footnote 6 on page 135: "Globally valid constraints could be saved and then re-used for other subproblems, but currently such feature is
not implemented." This is not great at all.

**Comment by odow** [comment]
For the near term, we can leave the example as-is. It builds and solves quickly on Julia 1.6, and we'll continue building the docs on 1.6 for as long as it is the LTS.

To close this issue, I think we just need to add a warning that GLPK has known performance issues with lazy constraints, and users are encouraged to use a commercial solver instead?

**Comment by jd-foster** [comment]
I don't know the example well enough, but why would the change from Julia 1.6 to 1.7 modify GLPK's performance?

**Comment by odow** [comment]
Two likely culprits:

 * The random numbers might be different
 * There were BLAS-related changes between Julia 1.6 and 1.7 (This might mean it's picking up different versions of dependencies? I didn't check though, so not sure on this one.)

**Comment by jd-foster** [comment]
Yeah, my first thought about this was your first item. See
https://julialang.org/blog/2021/11/julia-1.7-highlights/#new_rng_reproducible_rng_in_tasks

**Comment by joaquimg** [comment]
You can call `rng = Random.MersenneTwister` explicitly as a random number generator in both 1.6 (where it is default) and 1.7 (not default).

**Comment by odow** [comment]
Closed by #2863
